modules/ximgproc/src/weighted_median_filter.cpp

```diff
@@ -0,0 +1,718 @@
+							updateBCB(BCB[gval],BCBf,BCBb,gval,((fval <= medianVal)<<1)-1);
+						}
+					}
+
+
+					// Delete leaving pixels into joint-histogram and BCB
+					rownum = y - r;
+					if(rownum >= 0)
+					{
+						int *inputImgPtr = I.ptr<int>(rownum);
+						int *guideImgPtr = F.ptr<int>(rownum);
+						uchar *maskPtr = mask.ptr<uchar>(rownum);
+
+						for(int j=downX;j<=upX;j++)
+						{
+							if(!maskPtr[j])continue;
+
+							fval = inputImgPtr[j];
+							curHist = H[fval];
+							gval = guideImgPtr[j];
+
+							curHist[gval]--;
+
+							// Maintain necklace table of joint-histogram
+							if(!curHist[gval] && gval)
+							{
+								int *curHf = Hf[fval];
+								int *curHb = Hb[fval];
+
+								int p1=curHb[gval],p2=curHf[gval];
+								curHf[p1]=p2;
+								curHb[p2]=p1;
+							}
+
+							// Maintain necklace table of BCB
+							updateBCB(BCB[gval],BCBf,BCBb,gval,-((fval <= medianVal)<<1)+1);
+						}
+					}
+			}
+		}
+
+		// Deallocate the memory
+		{
+			delete []BCB;
+			delete []BCBf;
+			delete []BCBb;
+			int2D_release(H);
+			int2D_release(Hf);
+			int2D_release(Hb);
+		}
+
+		// end of the function
+		return outImg;
+	}
+}
+
+namespace cv
+{
+namespace ximgproc
+{
+	void weightedMedianFilter(InputArray joint, InputArray src, OutputArray dst, int r, double sigma, std::string weightType, Mat mask)
+	{
+		CV_Assert(!src.empty());
+		CV_Assert(r > 0 && sigma > 0);
+
+		int nI = 256;
+		int nF = 256;
+
+		Mat I = src.getMat();
+		Mat F = joint.getMat();
+		if(joint.empty())
+		{
+			medianBlur(src, dst, r);
+			return;
+		}
+
+		CV_Assert(I.depth() == CV_32F || I.depth() == CV_8U);
+		CV_Assert(F.depth() == CV_8U && (F.channels() == 1 || F.channels() == 3));
+
+		dst.create(src.size(), src.type());
+		Mat D = dst.getMat();
+
+		if(D.data == F.data)
+			F = F.clone();
+		if(D.data == I.data)
+			I = I.clone();
+
+		//Preprocess I
+		//OUTPUT OF THIS STEP: Is, iMap
+		//If I is floating point image, "adaptive quantization" is done in from32FTo32S.
+		//The mapping of floating value to integer value is stored in iMap (for each channel).
+		//"Is" stores each channel of "I". The channels are converted to CV_32S type after this step.
+		vector<float *> iMap(I.channels());
+		vector<Mat> Is;
+		split(I,Is);
+		for(int i=0;i<(int)Is.size();i++)
+		{
+			if(I.depth() == CV_32F)
+			{
+				iMap[i] = new float[nI];
+				from32FTo32S(Is[i],Is[i],nI,iMap[i]);
+			}
+			else if(I.depth() == CV_8U)
+			{
+				Is[i].convertTo(Is[i],CV_32S);
+			}
+		}
+
+		//Preprocess F
+		//OUTPUT OF THIS STEP: F(new), wMap
+		//If "F" is 3-channel image, "clustering feature image" is done in featureIndexing.
+		//If "F" is 1-channel image, featureIndexing only does a type-casting on "F".
+		//The output "F" is CV_32S type, containing indexes of feature values.
+		//"wMap" is a 2D array that defines the distance between each pair of feature indexes.
+		// wMap[i][j] is the weight between feature index "i" and "j".
+		float **wMap;
```

Comment:
It will be easier to avoid such errors with [cv::AutoBuffer](http://docs.opencv.org/master/d8/dd0/classcv_1_1AutoBuffer.html) for 1D arrays and cv::Mat for 2D arrays.